### PR TITLE
Add optional param for AWS IAM role assumption

### LINKED ISSUE #142: Allow specifying AWS IAM role to assume with IAM user keys
**Is your feature request related to a problem? Please describe.**

Currently drone-cache support AWS users, but often the permissions are granted to roles instead of users. The usage of roles also allows universally working API access when some policies are IP restricted as described here https://aws.amazon.com/premiumsupport/knowledge-center/iam-restrict-calls-ip-addresses/ 

**Describe the solution you'd like**
Add parameter to specify role ARN to assume, use access keys to assume role with STS and retrieve credentials with role rights.

### PR DESCRIPTION
**Pull request by hhamalai**
Fixes #142

## Proposed Changes
- add flag to specify role arn
- if role arn is specified, use access keys to assume the role with arn provided

### Description
Allows specifying role ARN which, if specified, is assumed with STS. All S3 operations are executed with this role.

### Checklist

- [x] Read the [**CONTRIBUTING**](CONTRIBUTING.md) document.
- [x] Read the [**CODE OF CONDUCT**](CODE_OF_CONDUCT.md) document.
- [x] Ensure your code follows the code style of this project.
- [x] Ensure CI and all other PR checks are green OR
    - [x] Code compiles correctly.
    - [x] Created tests which fail without the change (if possible).
    - [x] All new and existing tests passed.
- [x] Add your changes to `Unreleased` section of [CHANGELOG](CHANGELOG.md).
- [x] Improve and update the [README](README.md) (if necessary).

### CODE REVIEW

File: storage/backend/s3/config.go [hunk 1]
```diff
@@ -15,6 +15,7 @@ type Config struct {
 	Encryption string // if not "", enables server-side encryption. valid values are: AES256, aws:kms.
 	Endpoint   string
 	Key        string
+	RoleArn    string // if "", do not assume IAM role i.e. use the IAM user
```
Comment (kakkoyun) [inline storage/backend/s3/config.go:18]: nit
```suggestion
	RoleArn    string // if "", do not assume IAM role i.e. use the IAM user.
```

**Review by kakkoyun — CHANGES_REQUESTED** [review]
Looking good. Thanks a lot for the contribution. 

Is there way to test thing with our setup? We have minio, it'd be good to have tests.

**Comment by hhamalai** [comment]
> Is there way to test thing with our setup? We have minio, it'd be good to have tests.

I didn't see any authentication related tests in the code, at least for AWS. What's your take on mocked unit tests, that would be plausible, but I can also check the minio support for role assumption as documented here  https://github.com/minio/minio/blob/master/docs/sts/assume-role.md

**Comment by kakkoyun** [comment]
@hhamalai 
We have rudimentary happy path tests in the form of high-level e2e tests in here https://github.com/meltwater/drone-cache/blob/c6fa14678b59d689c99f28655880031a80911be7/internal/plugin/plugin_test.go#L386

What we can either extend those or even better add a specific case in here https://github.com/meltwater/drone-cache/blob/master/storage/backend/s3/s3_test.go

Even a simple one to prevent regressions would suffice.

File: storage/backend/s3/config.go (hunk 1 above)
Comment (hhamalai) [reply storage/backend/s3/config.go]: fixed

**Comment by hhamalai** [comment]
Hi, I added the test case, unfortunately it required quite a bit of too many changes. I am still not too familiar with minio, so take the following findings with a grain of salt:

1. need newer version minio in order to support STS operations
2. need a creation of users an policies on minio as the admin user cannot use STS (in docker-compose & drone yaml)

**Review by kakkoyun — APPROVED** [review]
Wow, what a great PR! Thank you very much for this contribution and level of code quality.

It looks good to me. I have only requested a single change concerning package imports. If you can fix it I'll merge this immediately.

**Review by kakkoyun — APPROVED** [review]

**Pull request merged**